### PR TITLE
proto: clean process_command()

### PR DESCRIPTION
**Pull request by dormando**
Pulled almost all of the inlined commands out of the function. Also
abstract out the token count checks.

Meta commands have a special fast handler

also reordered/restructured the rest of the commands. cleaner code and
should be same/similar performance for the common commands.

Need to do a quick verification bench and wait a bit to see if I still like the token macros later.

edit: also need to finish TLS mode

### CODE REVIEW

File: memcached.c [hunk 1]
```diff
@@ -2914,6 +2914,33 @@ typedef struct token_s {
...
+        if (min != -1 && ntokens < min) { \
+            out_string(c, "CLIENT_ERROR wrong number of command tokens"); \
+            return; \
+        } else if (max != -1 && ntokens > max) { \
+            out_string(c, "CLIENT_ERROR wrong number of command tokens"); \
+            return; \
+        } \
+    } while (0)
+
+#define WANT_TOKENS_OR(ntokens, a, b) \
```
Comment (minkikim89) [inline memcached.c]: we need this macro now?
Anything implemented with this macro can be replaced with WANT_TOKENS.
For example, WANT_TOKENS_OR(ntokens, 6, 7) == WANT_TOKENS(ntokens, 6, 7)

Comment (dormando) [reply memcached.c]: those're all new. in my first pass I had everything with just WANT_TOKENS(), but most of the paths were originally just two ||'s or a single >= so I added the extra two macros to not add significant slowdown.

File: memcached.c [hunk 2]
```diff
@@ -5750,321 +6019,162 @@ static void process_command(conn *c, char *command) {
...
+                conn_set_state(c, conn_mwrite);
+                break;
+            case 'a':
+                process_marithmetic_command(c, tokens, ntokens);
+                break;
+            case 'e':
+                process_meta_command(c, tokens, ntokens);
+                break;
+        }
+    } else if (tokens[COMMAND_TOKEN].value[0] == 'g') {
```
Comment (minkikim89) [inline memcached.c]: if input string is just a space(" "), segfault occurs here.
Please check it.

Comment (dormando) [reply memcached.c]: good catch! I left out the default: there. missing coverage in t/metaget.t

**Comment by dormando** [comment]
purposefully not putting this into 1.6.1. been glancing at it in case I want to rip out more of the token checks or something.

**Comment by dormando** [comment]
performance sucks, apparently. The ntokens shortcutting did a good job of keeping the strcmp's down in the old code.

Going to restructure it a bit more aggressively since we've got an "it might add up in latency" level slowdown for the non-get commands. Should also find a solution for the resp memset being too large since that's showing up under perf significantly.

**Comment by dormando** [comment]
to test I took the "touch" command and ran mc-crusher against an empty cache. Touch is halfway down the if/else tree. I compared `next` and this branch. For `next` I also moved the metaget commands to the bottom of the tree, since adding them had slowed things down a bit already.

Two mc threads vs 4 memcached worker threads. misses so I could see the overhead of the parse tree more clearly. ~600k misses/sec for modified `next` and ~700k for this branch after the fix I just pushed. So now it should be faster than it ever was for the common commands.

**Comment by dormando** [comment]
happy enough with this for now. going to merge it into `next` soon.

**Comment by dormando** [comment]
released in 1.6.3

**Pull request closed without merging**